---
title: "Validating_AFTOL"
author: "Michael Mann"
date: "3/31/2022"
output: html_document
---



```{r setup, include=FALSE}
library(tidyverse)
library(microseq)

```
This will be based on the compute phylogenetic taxonomy code i wrote for the pacbio project.

running raxml on each family. 


```{r read in data}



genomes_metadata <- read_csv("genomes_metadata.csv")
SSU_Alignment <- readFasta("")

AFTOL_Aligned <- 
  SSU_Alignment %>%
  filter(str_detect(string = Header, pattern = "Phylogenomic_", negate = T))


```


```{r custom functions}



# making this folder for everything. 
dir.create("AFTOL/Validation/")
dir.create("AFTOL/Validation/Region_18S")


### CREATE RAXML function!!
raxml <- function(alignment, path){
  
  paste0(
    "/Users/michaelmann/standard-RAxML-master/RAxML-8.0.3 -T 4 -n result -s ", 
    alignment, 
    " -m GTRCATIX -r /Users/michaelmann/Dropbox/Masters_Program/Genomes_rRNA/AFTOL/AFTOL_tree.tre -c 25 -p 23421 -f a -N 100 -x 23423 -o   Cryptosporidium_parvum -w ", 
    path) %>%
    system()

}



```


```{r processsing}

r18S <- 
  genomes_metadata %>%
  mutate(Header = paste0("Phylogenomic_", old_taxonID_linked_genome_sequence)) %>%
  mutate(Header = str_replace_all(Header, pattern = "'", replacement = "")) %>% # some of the strain species names had extra symbols. I removed them.
  mutate(Header = str_replace_all(Header, pattern = "-", replacement = "")) %>% 
  right_join(SSU_Alignment, by = "Header")  %>%
  select(Header, NCBI_Family, Sequence) %>%
  drop_na(Sequence) %>%
  replace_na(list(NCBI_Family = "Outgroup")) %>%
  group_by(NCBI_Family) %>%
  nest() %>% 
  ungroup() %>%
  mutate(AFTOL = list(AFTOL_Aligned)) %>%
  mutate(merged = map2(data, AFTOL, bind_rows)) %>%
  mutate(path = paste0("AFTOL/Validation/Region_18S/", NCBI_Family)) %>%
  select(-data, -AFTOL) %>%
  mutate(alignment_path = paste0(path, "/", NCBI_Family, "_aligned.fasta")) %>%

  
  # write alignments with AFTOL and one family to file
r18S %>% 
  walk2(.x = merged, .y = alignment_path,  .f  = writeFasta(fdta = .x, out.file = .y))

# run raxmly on each family
r18S %>%
  walk2(.x = alignment_path, .y = .$path, .f = raxml(alignment = .x, path = .y)) #create the folders for each family




# run raxml for each. 
r18S %>%
  walk2(.x = 
    
    .l = list(.$path_unaligned, .$path, .$NCBI_Family), .f = MAFFT)


  
```
